### DISCUSSION TITLE
How can I locally debug the code of the workflow server? 🙏

### DISCUSSION THREAD

**Discussion by mio4kon** [argoproj/argo-workflows#12587] [Q&A] [answered]
I am currently using the 'Developing locally' approach with the k3d method. However, I can only debug by logging at the moment, and I have to rebuild it every time. Is there a good tutorial on how to use Golang for debugging the workflow-controller?"

I am currently following the approach outlined in https://argo-workflows.readthedocs.io/en/latest/running-locally/.

**Comment by jswxstw** [comment, marked as answer]
Yes, you can try it as below:
<img width="1005" alt="image" src="https://github.com/argoproj/argo-workflows/assets/18543532/e9c21f9a-ddf7-4aef-890e-3903b92ff24f">

Environment example:
```
ARCHIVED_WORKFLOW_GC_PERIOD=30s;ARGO_EXECUTOR_PLUGINS=false;ARGO_LOG_LEVEL=debug;ARGO_MANAGED_NAMESPACE=argo;ARGO_NAMESPACE=argo;ARGO_NAMESPACED=true;ARGO_PROGRESS_PATCH_TICK_DURATION=7s;ARGO_REMOVE_PVC_PROTECTION_FINALIZER=true;DEFAULT_REQUEUE_TIME=1s;LEADER_ELECTION_DISABLE=true;MAX_OPERATION_TIME=3600s;UPPERIO_DB_DEBUG=1;WORKFLOW_GC_PERIOD=30s
```
`LEADER_ELECTION_DISABLE` should be true and `MAX_OPERATION_TIME` should be increased in debug mode, other envs can be modified as needed.

Go Tool arguments:
```
-v -ldflags '-X github.com/argoproj/argo-workflows/v3.version=latest'
```

By the way, you can use `kit ui` to run argo server and ui without workflow-controller locally.

**Reply by mio4kon** [reply]
Thanks so much! I'll give it a try

**Reply by mio4kon** [reply]
@jswxstw hi，What does 'Kit UI' mean? 
Right now, when I run `make start API=true` or `make start UI=true`, I get an error because the controller is already started for debugging. Are there any other commands to start the UI and Argo server separately?

**Reply by mio4kon** [reply]
I got it, 'Kit UI' is a command, haha! 🙏

**Reply by jswxstw** [reply]
https://github.com/argoproj/argo-workflows/blob/42262690fcbab0ed6106d3f9069e128655b3339f/Makefile#L561-L563
Yes, `make start` => `kit ${TASKS}`
